### PR TITLE
AsyncIOProcessor preserve thread context

### PR DESCRIPTION
**Pull request by henningandersen**
AsyncIOProcessor now preserves thread context, ensuring that deprecation
warnings are not duplicated to other concurrent operations on the same
shard.

### CODE REVIEW

**Review by s1monw — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/common/util/concurrent/AsyncIOProcessor.java [hunk 1]
```diff
@@ -121,6 +123,16 @@ private void processList(List<Tuple<Item, Consumer<Exception>>> candidates) {
         }
     }
 
+    private Consumer<Exception> preserveContext(Consumer<Exception> consumer) {
+        ThreadContext.StoredContext storedContext = threadContext.newStoredContext(false);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/AsyncIOProcessor.java]: perhaps use `threadContext.newRestorableContext(false)`. It's simpler.

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by henningandersen** [comment]
@elasticmachine run elasticsearch-ci/default-distro

File: server/src/main/java/org/elasticsearch/common/util/concurrent/AsyncIOProcessor.java (hunk 1 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/AsyncIOProcessor.java]: Thanks, fixed here: acfa0d6880a2a5a0ba8eacc07a02172459152c92

**Pull request merged**